src/packagedcode/golang.py

```diff
@@ -0,0 +1,144 @@
+from packagedcode import models
+
+
+"""
+Handle Go packages including go.mod and go.sum files.
+"""
+"""
+Sample go.mod file:
+module github.com/alecthomas/participle
+require (
+	github.com/alecthomas/repr v0.0.0-20181024024818-d37bc2a10ba1
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/stretchr/testify v1.4.0
+)
+go 1.13
+"""
+
+# TODO:
+# go.mod file does not contain version number.
+# valid download url need version number
+# CHECK: https://forum.golangbridge.org/t/url-to-download-package/19811
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class GolangPackage(models.Package):
+    metafiles = ('go.mod',)
+    default_type = 'golang'
+    default_primary_language = 'Go'
+    default_web_baseurl = 'https://pkg.go.dev'
+    default_download_baseurl = None
+    default_api_baseurl = None
+
+    @classmethod
+    def recognize(cls, location):
+        if fileutils.file_name(location).lower() == 'go.mod':
+            gomod_obj = GoMod()
+            gomod_data = gomod_obj.parse_gomod(location)
+            yield build_gomod_package(gomod_data)
+
+    @classmethod
+    def get_package_root(cls, manifest_resource, codebase):
+        return manifest_resource.parent(codebase)
+
+    def repository_homepage_url(self):
+        return self.homepage_url
+
+
+def build_gomod_package(gomod_data):
+    """
+    Return a Package object from a go.mod file or None.
+    """
+    package_dependencies = []
+    require = gomod_data.get('require') or []
+    for namespace, name, version in require:
+        package_dependencies.append(
+            models.DependentPackage(
+                purl=PackageURL(
+                    type='golang',
+                    namespace=namespace,
+                    name=name
+                ).to_string(),
+                requirement=version,
+                scope='require',
+                is_runtime=True,
+                is_optional=False,
+                is_resolved=False,
+            )
+        )
+
+    exclude = gomod_data.get('exclude') or []
+    for namespace, name, version in exclude:
+        package_dependencies.append(
+            models.DependentPackage(
+                purl=PackageURL(
+                    type='golang',
+                    namespace=namespace,
+                    name=name
+                ).to_string(),
+                requirement=version,
+                scope='exclude',
+                is_runtime=True,
+                is_optional=False,
+                is_resolved=False,
+            )
+        )
+
+    name = gomod_data.get('name')
+    namespace = gomod_data.get('namespace')
+    homepage_url = 'https://pkg.go.dev/{}'.format(gomod_data.get('module'))
+
+    return GolangPackage(
+        name=name,
+        namespace=namespace,
+        homepage_url=homepage_url,
+        dependencies=package_dependencies
+    )
```

Comment:
with `module github.com/alecthomas/kingpin` would you be able to infer some better homepage URL instead? 
And what you have in homepage_url would be instead best computed as ` repository_download_url` in https://github.com/nexB/scancode-toolkit/blob/3068680e822bb6ed58d784ecb36ad901ee158078/src/packagedcode/models.py#L250 

You need not just copy it over from homepage_url which would be incorrect but instead compute it in that method somewhere along the lines of `'https://pkg.go.dev/{namespace}/{name}@{version}/{subpath}'.format(....)`

For instance https://pkg.go.dev/github.com/stretchr/testify@v1.6.1/mock

Also since this is obviously from Github, what about also filling in the vcs_url? https://github.com/nexB/scancode-toolkit/blob/3068680e822bb6ed58d784ecb36ad901ee158078/src/packagedcode/models.py#L429

(Note that I would be fine that you handle that in a new ticket, in which case enter a ticket for that)

Reply:
I am creating a new ticket for this :)

Reply:
Added new ticket for this https://github.com/nexB/scancode-toolkit/issues/2121